File: src/main/java/fi/quanfoxes/parser/UnresolvedType.java
package fi.quanfoxes.parser;

import fi.quanfoxes.parser.nodes.TypeNode;
import fi.quanfoxes.parser.nodes.UnresolvedIdentifier;

public class UnresolvedType extends Type implements Resolvable {
    private Resolvable resolvable;

    public UnresolvedType(Context context, String name) {
        super(context);
        this.resolvable = new UnresolvedIdentifier(name);
    }

    public UnresolvedType(Context context, Resolvable resolvable) throws Exception {
        super(context);
        this.resolvable = resolvable;
    }

    @Override
    public Node resolve(Context context) throws Exception {
        Node resolved = resolvable.resolve(context);

        if (resolved instanceof Contextable) {
            Contextable contextable = (Contextable)resolved;
            return new TypeNode((Type)contextable.getContext());
        }

        throw new Exception("Couldn't resolve type");
    }
}

File: Zigzag/Parser/UnresolvedType.cs
public class UnresolvedType : Type, IResolvable
{
	private IResolvable Resolvable { get; }

	public UnresolvedType(Context context, string name) : base(context)
	{
		Resolvable = new UnresolvedIdentifier(name);
	}

	public UnresolvedType(Context context, IResolvable resolvable) : base(context)
	{
		Resolvable = resolvable;
	}

	public Node? Resolve(Context context)
	{
		if (Resolvable.Resolve(context) is IType resolved && resolved.GetType() != null)
		{
			return new TypeNode(resolved.GetType()!);
		}

		return null;
	}

	public Type? TryResolveType(Context context)
	{
		return Resolve(context)?.GetType();
	}

	public Status GetStatus()
	{
		return Resolvable.GetStatus();
	}
}